language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -166,59 +315,87 @@ instantiateEvaluate con ty fun
 -- If succesful, proceed with either this same term or with the result of the computation
 -- depending on whether 'BuiltinName' is saturated or not.
 applyEvaluate
-    :: VarEnv -> VarEnv -> Context -> Plain Value -> Plain Value -> CekM EvaluationResultDef
+    :: VarEnv -> VarEnv -> Context -> WithMemory Value -> WithMemory Value -> CekM w s EvaluationResultDef
 applyEvaluate funVarEnv argVarEnv con (LamAbs _ name _ body) arg =
     withVarEnv (extendVarEnv name arg argVarEnv funVarEnv) $ computeCek con body
 applyEvaluate funVarEnv _         con fun                    arg =
-    let term = Apply () fun arg in
+    let term = Apply 1 fun arg in -- TODO
         case termAsPrimIterApp term of
             Nothing                       ->
-                throwError $ MachineException NonPrimitiveApplicationMachineError term
+                throwError $ MachineException NonPrimitiveApplicationMachineError (void term)
             Just (IterApp headName spine) -> do
-                constAppResult <- applyStagedBuiltinName headName spine
+                constAppResult <- applyStagedBuiltinName arg headName spine
                 withVarEnv funVarEnv $ case constAppResult of
                     ConstAppSuccess res -> computeCek con res
                     ConstAppFailure     -> pure EvaluationFailure
                     ConstAppStuck       -> returnCek con term
                     ConstAppError   err ->
-                        throwError $ MachineException (ConstAppMachineError err) term
+                        throwError $ MachineException (ConstAppMachineError err) (void term)
+
+evaluateInCekM :: EvaluateConstApp (ExceptT CekMachineException (RWS CekEnv w s)) a -> CekM w s (ConstAppResult a)
+evaluateInCekM a = do
+    let eval means' term = local (over cekEnvMeans (mappend means')) $ computeCek [] (annotateMemory term)
+    runEvaluateConstApp eval a
 
-evaluateInCekM :: EvaluateConstApp (Either CekMachineException) a -> CekM (ConstAppResult a)
-evaluateInCekM a =
-    ReaderT $ \cekEnv ->
-        let eval means' = evaluateCekIn $ cekEnv & cekEnvMeans %~ mappend means'
-            in runEvaluateConstApp eval a
+-- TODO invalid number of args
+estimateStaticStagedCost :: BuiltinName -> [WithMemory Value] -> (ExCPU, ExMemory)
+estimateStaticStagedCost AddInteger [x, y] = (1, 1)
```

Comment:
That should probably be a part of [`DynamicBuiltinNameMeaning`](https://github.com/input-output-hk/plutus/blob/master/language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs#L112) somehow.

Reply:
Yes, I definitely agree with this. Except presumably the static bits can be static as their meanings are currently static. But the cost functions for dynamic builtins should be provided at runtime.

Reply:
I'll table this for now, it's still a TODO. Will be solved by a later PR.